#!/usr/bin/env python3
# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.

import random
import unittest

import numpy as np
import torch
from ml.rl.test.gridworld.gridworld_base import DISCOUNT
from ml.rl.test.gridworld.gridworld_continuous import GridworldContinuous
from ml.rl.test.gridworld.gridworld_evaluator import GridworldDDPGEvaluator
from ml.rl.test.gridworld.gridworld_test_base import GridworldTestBase
from ml.rl.thrift.core.ttypes import (
    DDPGModelParameters,
    DDPGNetworkParameters,
    DDPGTrainingParameters,
    RLParameters,
)
from ml.rl.training.ddpg_trainer import DDPGTrainer


class TestGridworldDdpg(GridworldTestBase):
    def setUp(self):
        self.minibatch_size = 4096
        super(self.__class__, self).setUp()
        np.random.seed(0)
        random.seed(0)
        torch.manual_seed(0)

    def get_ddpg_parameters(self):
        return DDPGModelParameters(
            rl=RLParameters(
                gamma=DISCOUNT,
                target_update_rate=0.1,
                reward_burnin=100,
                maxq_learning=True,
            ),
            shared_training=DDPGTrainingParameters(
                minibatch_size=self.minibatch_size,
                final_layer_init=0.003,
                optimizer="ADAM",
            ),
            actor_training=DDPGNetworkParameters(
                layers=[-1, 256, 128, -1],
                activations=["relu", "relu", "tanh"],
                learning_rate=0.05,
                l2_decay=0.01,
            ),
            critic_training=DDPGNetworkParameters(
                layers=[-1, 256, 256, 128, -1],
                activations=["relu", "relu", "relu", "linear"],
                learning_rate=0.05,
                l2_decay=0.01,
            ),
        )

    def _test_ddpg_trainer(self, use_gpu=False, use_all_avail_gpus=False):
        self.check_tolerance = False
        self.tolerance_threshold = 1.0
        environment = GridworldContinuous()
        trainer = DDPGTrainer(
            self.get_ddpg_parameters(),
            environment.normalization,
            environment.normalization_action,
            environment.min_action_range,
            environment.max_action_range,
            use_gpu=use_gpu,
            use_all_avail_gpus=use_all_avail_gpus,
        )
        evaluator = GridworldDDPGEvaluator(environment, DISCOUNT)
        self.evaluate_gridworld(environment, evaluator, trainer, trainer, use_gpu)

    def test_ddpg_trainer(self):
        self._test_ddpg_trainer()

    @unittest.skipIf(not torch.cuda.is_available(), "CUDA not available")
    def test_ddpg_trainer_gpu(self):
        self._test_ddpg_trainer(use_gpu=True)

    @unittest.skipIf(not torch.cuda.is_available(), "CUDA not available")
    def test_ddpg_trainer_all_gpus(self):
        self._test_ddpg_trainer(use_gpu=True, use_all_avail_gpus=True)
